Workbench - daskify tripyview - new version

patrickscholz opened the pull request.

patrickscholz committed:
delete print stetement

patrickscholz committed:
add routines for initialising notebook header and for the shortcute for dask client initialisation, adding spinupcycles index to input paths ...

patrickscholz committed:
improving fixing hovmöller diagramms, chunk data loading for hovmöller computation in time spice

patrickscholz committed:
improve notebook to compute horizontal slices, create dask parallelized horizontal coarse graining alogorythm, chunk horizontal plooting for pcolor plot,

patrickscholz committed:
fix quiver computation for larger grid like dart mesh, nevertheless there are still to many arrows in quiver plot at large grids

patrickscholz committed:
improve transect computation. make transect analyser much faster, improve transport and heatflux computation through transects now project transport on normal vector of main transect

patrickscholz committed:
change input from data loading routine from data_vecrot to data_rot which is more universal across the package

patrickscholz committed:
improve notebook headers in 2dmesh and 3d sphere routine

patrickscholz committed:
add subroutine to centralize and unify the notebook header, which figure environment if  inline, notebook or widget can be chossen by input

patrickscholz committed:
define several shortcut routines: to load the dask client, to add the spinupcycle number to the inputpath ...

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
update notebook without error message

patrickscholz committed:
update notebook without error message

patrickscholz committed:
improve zonal mean routine, add meridional mean routine, fix clabel issue when reference plot, add loading of bolus velocity when doing streaml

patrickscholz committed:
also adapt tripyrun drivers for the new meridionally averaged transects

patrickscholz committed:
update README file

patrickscholz committed:
update zmoc zmean mmean ... computation,

patrickscholz committed:
further update the entire code, towards parallelisation with dask, close memory leakages and minimize dask task graph

patrickscholz committed:
further improve trripyview towrds dask

patrickscholz committed:
fix scalar transect routine with respect to dask usage

patrickscholz committed:
finish the actualization of all the transect routiens with respect to dask optimisation

patrickscholz committed:
attempt to fix/improve ,eridional heatflux computation routine

patrickscholz committed:
fix meridional heatflux notebooks for small grids

patrickscholz committed:
fix small bug

patrickscholz committed:
fix a couple of small problems that emerged with AO04 mesh, especially with the land sea mask patch

patrickscholz committed:
some changes made to ensure functionallity on juwels

patrickscholz committed:
small changes due juwels

patrickscholz committed:
small change in hflx routien

patrickscholz committed:
try to improve heatflux computation

patrickscholz committed:
update and improve heatflux computations

patrickscholz committed:
start new hbarstreamf routine

patrickscholz committed:
improve streamfunction computation with dask

patrickscholz committed:
update

patrickscholz committed:
Merge branch 'main' into workbench

patrickscholz committed:
make mhflx computation work for dart mesh, there number of dask workers needd to be reduced otherwise ones permanently runs into netcdf/HDF5 problems

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
try to fix a couple of issues when computing dmoc for very large grids >10M vertices

patrickscholz committed:
improved dmoc computations

patrickscholz committed:
improved dmoc computations

patrickscholz committed:
imrove dmoc computation for large grids

patrickscholz committed:
add gradient computation capabilities

patrickscholz committed:
further improvments

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
small improvments to gradient computation, and matplotlib streamline plotting

patrickscholz committed:
apply minor major improvements, especially kickout compat option from engine_dict --> creates big problems

patrickscholz committed:
kickout another option from netcdf engine_dict

patrickscholz committed:
fix small issue in streamline plotting when working with chunked data

patrickscholz committed:
fix a couple of minor issues

patrickscholz committed:
fix new error messsage

patrickscholz committed:
be able to switch off the usage of .sub_3dsphere with an environment variable

patrickscholz committed:
be able to use monthly and seasonal PHC climatology

patrickscholz committed:
make some minor adjustments to the plotting routine, make show there is a plt.show() at every end, make sure that flushing evetns are only when there are more chunks ...

patrickscholz committed:
make sure that transect position figures can be saved

patrickscholz committed:
add changes from paul to tripyrun, notebook and figures paths can be yaml file inputparameter

patrickscholz committed:
fix problem when lsmask polygons become a GeometryCollection after make_valid(poly)

patrickscholz committed:
add functionality to add climatoligical reference line in tline plot and reference contourline in hslice together with a contour line from a climatological lon/lat field

patrickscholz committed:
fix small bug when plotting horizontal colorbar, and small problem for the 3d sphere whenn all boundary indices are equal 1

patrickscholz committed:
change from record=-1 to record=None

patrickscholz committed:
fix back template_hslice.ipynb

patrickscholz committed:
check if tri has attribute mask_n_box or mask_e_box

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
add script to do vector rotation and interpolation from e2n offline

patrickscholz committed:
add some comments to tools/do_rot_ie2n.ipynb

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
fix the issue, so that e2n interpolation can be done on vertically selected data

patrickscholz committed:
change output string

patrickscholz committed:
update ../../tools/do_rot_ie2n.ipynb

patrickscholz committed:
make ure there is a grid_rotmat_dask(abg) and grid_rotmat(abg) routine

patrickscholz committed:
small change in the print output

patrickscholz committed:
make sure there is a grid_rotmat_dask(abg) amd grid_rotmat_(abg) routine

patrickscholz committed:
Update template_transp_dmoc.ipynb

fix small bracket issue in notebook

patrickscholz committed:
Update template_transp_dmoc.ipynb

patrickscholz committed:
fix small bug in computing the weight for nz

patrickscholz committed:
make sure that None in the data_list work as an empty placeholder

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
switch off default anomaly plots

patrickscholz committed:
update templates_notebooks/template_hslice.ipynb for depth=bottom, bottom-1 bottom-2

patrickscholz committed:
update templates_notebooks/template_hslice.ipynb for depth=bottom, bottom-1 bottom-2

patrickscholz committed:
small issue in parallel computation of transport

patrickscholz committed:
improve categorical cmap

patrickscholz committed:
add template for absolute and relative temperature/salinity heatmap

patrickscholz committed:
Update sub_transect.py

small indices failure for transport computation when velocities are on nodes

patrickscholz committed:
change default to flat shading better accurate representation of the x-direction points

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
switch seawater package to gsw package to compute potential temperature, density ...

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
change comment

patrickscholz committed:
fix and improve gradient computation based on the meshdiag file

patrickscholz committed:
small bugfix in grid_interp_e2n

patrickscholz committed:
small improvements

patrickscholz committed:
improve gradient operation and ie2n operation

patrickscholz committed:
strongly speedup ie2n operation for 1d, 2d, and 3d up to a factor of x200 by using numba jit

patrickscholz committed:
extend numba for computing boundary edges, nINe, eINe and nINn arrays

patrickscholz committed:
improve vector rotation routines to jit numba, brings some speed up

patrickscholz committed:
change some filenames, reactivate pickle but made joblib the default

patrickscholz committed:
roll back pickle library use

patrickscholz committed:
further optimize mesh operations make things generally faster

patrickscholz committed:
fix some stuff when loading e_resol, n_resol ...

alxthor committed:
Fix typo

patrickscholz committed:
push further the numba-fication of the mesh loading routine, numba-ficate and optimize the mesh reading, the lsmask computation the finding and augmentation of the periodic boundary ...

patrickscholz committed:
Merge pull request #214 from alxthor/workbench

Fix typo

patrickscholz committed:
add numba kernels for vertice and element smoothing, add these kernels via wrappers to the meshobject, do numba warmup for these kernels, kick out old routine from utility.py

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
add improved routine to compute periodic boudnary augmented Land Sea mask

patrickscholz committed:
fix small issue when choosen depth is smaller than the first layer depth

patrickscholz committed:
add neverworld2 functionality

patrickscholz committed:
improve some comments

patrickscholz committed:
add Neverworld plotting to tripyview

patrickscholz committed:
change README

patrickscholz committed:
add neverworld example plot

patrickscholz committed:
update REadme

patrickscholz committed:
update image

patrickscholz committed:
update image

patrickscholz committed:
add new faster routine to compute which edges are cutted by the transect line

patrickscholz committed:
update all notebooks

patrickscholz committed:
fix small bug in template_hmesh

patrickscholz committed:
some minor updates

patrickscholz committed:
overhaul part 1 of all notebooks with optimized chunking strategy

patrickscholz committed:
overhaul part 2 of all notebooks with optimized chunking strategy

patrickscholz committed:
make dmitry sein format usable

patrickscholz committed:
fix small dimension bug in vertical interpolation

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
expand file patterns, make dmitry seins data usable

patrickscholz committed:
fix problem when CFTime index

patrickscholz committed:
fix wrong vector routine, fix transect when transect segment does not cut any edges

patrickscholz committed:
fix horbarstreamf

patrickscholz committed:
fix horbarstreamf

patrickscholz committed:
overhaul part3

patrickscholz committed:
overhaul part 4

patrickscholz committed:
...

patrickscholz committed:
exclude lev_bounds

patrickscholz committed:
fix contour algorithm for special case when a boundary node has 4 neighbors

patrickscholz committed:
improve reg quiver plot a bit

patrickscholz committed:
improve reg quiver plot a bit

patrickscholz committed:
make rolling/running mean workable

JanStreffing commented:
This appears to be a major upgrade. As we are looking towards making tripyview an automatic post-processing for any fesom run done with esm_tools (also where fesom is a component model in coupled setups). Would it be feasible to pick this up again?

patrickscholz commented:
@JanStreffing: In principle yes, this branch is supposed to be the next release version for tripyview. Im still tinkering here and there a bit around as my own alpha tester whenever i encounter some issues while using it. This version supposedly should work a bit better for large meshes because now it is fully dask parallelized before it was a mixture of dask and joblib. I also quite substantially overhauled the reading and processing of the the mesh object, as well as the 1d,2d,3d vector rotation and elem2node interpolation  with numba, which also had quite a performance gain.

patrickscholz committed:
add commit from paul gierz about putting the relative path instead of absolute path of the fesom_logo.png

patrickscholz committed:
improve quiver plot, martin losch found issue in cartopy transform_vector routine in polar projection, now we do an own and point method to do this vector rotation into the projection by our self

patrickscholz committed:
improve quiver plot

patrickscholz committed:
add option for not finishing the figure with plt.show, allows to add more features into the plot

patrickscholz committed:
add glorys uv climatology to quiver notebook template

patrickscholz committed:
add some error message

patrickscholz committed:
improve computation of mesh resolution for elements and nodes

alxthor committed:
tripyrun/papermill: Ignore TemplateNotebook Kernel

Use default Kernel "python3" instead.

This should fix issues like `Unexpected Error: No such kernel named
py39_new` when running tripyrun.
Error was raised as papermill by default tries executing the notebooks
using the kernel specified in the notebook metadata.

This fix tells papermill to use the Kernel called "python3" and ignore
notebook metadata.

Details:
Once any conda environment is activated it seems that the kernel
"python3" is associated with the activated conda environment independent
of the name of the conda environment.

Only if no environment is activated "python3" is the name of the kernel
associated with the default e.g. Levante python env.

This can be verified by running `jupyter kernelspec list`

Only tested on Levante with python 3.12

patrickscholz committed:
Merge pull request #217 from alxthor/workbench

tripyrun/papermill: Ignore TemplateNotebook Kernel

alxthor committed:
Update all Notebooks to Suppress Tripyrun Parsing Error

papermill does not seem to like "=" signs in comments

It would come up with the error:

```
Unable to parse line 6 'parallel_nthread  = 2    # number of threads per worker, i.e. number worker = parallel_nprc/parallel_nthread'.
```

patrickscholz committed:
Merge pull request #218 from alxthor/workbench

Update all Notebooks to Suppress Tripyrun Parsing Error

alxthor committed:
Fix ReadOnly Error when Loading Climatology

Newer environments (e.g. python 3.12) would raise error about lon coordinate of climatology being read only when trying to roll across +-180 degrees, (i.e. lon.flags.writeable was False)
This is likely because of changes in dependencies and whether views or copies are returned.

Now always force copy.

Original Error Message:
```
ValueError                                Traceback (most recent call last)
Cell In[6], line 11
      7 if (vname in ['temp', 'sst', 'salt', 'sss', 'pdens'] or 'sigma' in vname) and (depth != 'bottom'):
      8     ts = clock.time()
      9     if   vname in ['temp', 'sst'] and  which_clim.lower()=='woa18': clim_vname = 't00an1'
     10     elif vname in ['salt', 'sss'] and  which_clim.lower()=='woa18': clim_vname = 's00an1'
---> 11     clim = tpv.load_climatology(mesh, clim_path, clim_vname, depth=depth)
     12     print(' --> elasped time to load clim: {:3.2f} min.'.format( (clock.time()-ts)/60  ))
     13     print(' --> clim uses {:3.2f} Gb:'.format(clim.nbytes/(1024**3)))
     14     print('')

File ~/tripyview/tripyview/sub_climatology.py:182, in load_climatology(mesh, datapath, vname, mon, depth, depidx, do_zarithm, do_hinterp, do_zinterp, descript, do_ptemp, pref, do_compute, do_load, do_persist, do_zweight, do_hweight, chunks, do_parallel, **kwargs)
    180     idx = np.where(lon>mesh.focus+180.0)[0]
    181     idx_roll = idx[0]
--> 182     lon[idx] = lon[idx]-360.0
    183 elif any(lon<mesh.focus-180.0):
    184     idx = np.where(lon<mesh.focus+180.0)[0]

ValueError: assignment destination is read-only
```

alxthor committed:
Fix deprecated time resampling

`AS` got changed into `YS` in recent pandas/xarray versions

alxthor committed:
Fix Warning about using cftime decoding in xarray

In newer environments (e.g. python 3.12) this warning would be issued by all workers:
```
FutureWarning: Usage of 'use_cftime' as a kwarg is deprecated. Please pass a 'CFDatetimeCoder' instance initialized with 'use_cftime' to the 'decode_times' kwarg instead.
Example usage:
time_coder = xr.coders.CFDatetimeCoder(use_cftime=True)
ds = xr.open_dataset(decode_times=time_coder)
```

patrickscholz committed:
add spectral dimension from idemix2 to plot along

patrickscholz committed:
add runid string to possible cmoc input

patrickscholz committed:
fix problem of dateline crossing transects

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

patrickscholz committed:
Merge pull request #219 from alxthor/workbench

Workbench

alxthor committed:
Transects running NE-->SW now do not crash anymore

A part of the code used to attempt to flip the transect direction if it
is running in that direction. It however assumed numpy arrays. Thus it
would crash on lists.

patrickscholz committed:
Merge pull request #222 from alxthor/workbench

Transects running NE-->SW now do not crash anymore

alxthor committed:
Fix problem with dateline crossing transects when plotting

During plot_vslice the x axis label is determined in
do_data_prepare_vslice by looking at the angle of the transect. This
decides whether lon/lat/dist is used. With dateline crossing transects
the angle determination was incorrect. Should hopefully be fixed now.

alxthor committed:
Mesh Resolution Template: Fix switch do_mesh

Remove trailing comma. This would lead `do_mesh` to be populated with an
non-empty tuple which evaluates to True

patrickscholz committed:
Merge pull request #223 from alxthor/workbench

Fix problem with dateline crossing transects when plotting

patrickscholz committed:
add a couple of more colormaps

patrickscholz committed:
add plt.show(block=False)

patrickscholz committed:
Merge branch 'workbench' of https://github.com/FESOM/tripyview into workbench

alxthor committed:
Small fixes and improvements to jupyter templates

template_hovm.ipynb: Fix case of plotting with reference data (Indexing and data cleanup). Note however that ref_year etc. do not do anything currently as plotting with reference data just does the anomaly of entire time series.

template_transp_hbstreamf.ipynb: Fix cleanup: 'data' has already been deleted and cannot be removed again

template_var_t.ipynb: Make script more robust. Now it does not crash anymore if the time period provided in year starts before data is available.

patrickscholz committed:
Merge pull request #224 from alxthor/workbench

Small fixes and improvements to jupyter templates

patrickscholz committed:
Fix 24 correctness bugs found in workbench audit (see BUGREPORT.md)

Covers crashes/silent corruption in dmoc, zhflx, tripyrun -d/-v CLI,
climatology comparisons, transect/zmoc/mesh diagnostics, and several
plotting/colorbar/driver-dispatch issues. Details and rationale for
each fix are in BUGREPORT.md.

patrickscholz committed:
Rework plot zorder layering and skip redundant setup in Dask workers

Move mesh/gridline/quiver/streamline/lsmask overlays from negative to
positive zorder values so layering is explicit and consistent instead
of relying on implicit ordering relative to the base data layer.

Add _running_in_dask_worker() (sub_utility.py) and use it to skip the
pkg_path print and numba kernel warm-up on Dask worker subprocesses,
since multiprocessing.current_process() reliably identifies a worker
even when distributed.get_worker() can't yet (spawn re-executes the
top-level module before the Worker object exists).

patrickscholz committed:
Support older xarray without xr.coders.CFDatetimeCoder in load_data_fesom2

xr.coders.CFDatetimeCoder was only added in xarray ~2024.09; environments
pinned to an older xarray (e.g. capped by Python 3.9, which xarray dropped
around the same release) don't have it. Fall back to passing use_cftime
directly to open_mfdataset in that case.

patrickscholz committed:
Fix ndens chunk-inconsistency crash in load_dmoc_data

The node->element transferred dmoc/dmoc_bolus variables were rechunked
along ndens either by reading chunksizes off the very dataset being
reassigned (crashes immediately, since it's already inconsistent) or
via a fresh 'auto' (no guarantee of matching the chunk boundaries the
natively dask-loaded ndens_h etc. already have in data_dMOC). Both now
take their ndens chunking from data_dMOC directly, so they align with
what's already there once merged.

patrickscholz committed:
Fix NaN-poisoning of dmoc srf/inner/fh/fw/fr fields in load_dmoc_data

FESOM's output stream converts exact zeros to NaN as part of its land-sea
masking convention. A flux term that is legitimately zero everywhere (e.g.
std_frwt_flux/std_rest_flux when a run has no salinity restoring or
freshwater forcing active) therefore comes back all-NaN, not all-zero.
Summing it straight in silently poisons the combined srf/inner/fh/fw/fr
transformation with NaN, which later skipna=True sum/cumsum calls render as
a flat zero -- indistinguishable from "nothing happening" instead of a
masked term. fillna(0) each flux field right after loading so a genuinely
absent term can't wipe out the others.

Also add None-guards after each load_data_fesom2 call (and at the top of
calc_dmoc_dask) so a missing variable/file returns None cleanly instead of
crashing on .rename()/attribute access.
